amm feature for the CRT pallet

dobertRowneySr opened the pull request:
Adds the AMM (also referred as Bonding Curve) feature to the project token pallet https://github.com/Joystream/joystream/issues/3630

# Features
- [X] `activate, deactivate, bond, unbond` extrinsics
- [X] slippage protection mechanisms: tolerance + deadline
- [X] user can bond and have its account immediately created with bloat bond deposited
- [X] linear pricing formula `ax + b` with, parameters arbirarely chosen by the user
- [ ] chose correct type for parameters `Permill` vs `Perquintill` (all references seems to point out that Permill accuracy should be sufficient)
- [x] added extrinsics to the content pallet for amm activation/deactivation w/ tests 
- [ ] Rug pulling protection mechanisms (~~to be discussed with bedeho?~~ part of the #4151 )
- [X] Doc lines above new extrinsics

# References
- For the choices of parameters `a = .003,b = 0` I used DeSo https://docs.deso.org/about-deso-chain/readme#the-creator-coin-supply-curve
- https://github.com/Joystream/joystream/issues/4406
- https://github.com/Joystream/joystream/issues/3754
- https://www.trality.com/blog/what-is-slippage-in-crypto (suggests that average slippage tolerance is typically in the 2-10% range)

# Computation summary (for the `eval` function)
Relation between JOY and CRT to:
## Bond
$$ \int_X^{X + x} (at + b) dt = \frac{a}{2} [(X + x)^2 - X^2] + b x $$
## Unbond
$$ \int_X^{X - x} (at + b) dt = - \bigg( \frac{a}{2} [X^2 - (X - x)^2] + b x  \bigg) = -\text{JOYs necessary for unbonding}$$
- $X$ = `token_info.total_supply` value before calling the extrinsic
- $x$ = `amount` parameter in the extrinsic

I have expanded the computation into 3 terms to avoid chances of overflowing for big numbers like $X^2$

dobertRowneySr committed:
chore: add unit test file for amm

dobertRowneySr committed:
chore: add test for unbonding & activation

dobertRowneySr committed:
Revert "chore: add public functions"

This reverts commit 862993afdb9838c52066cdfe099444c863f8bbfc.

dobertRowneySr committed:
chore: add tests on activate amm

dobertRowneySr committed:
fix: cargo fmt

dobertRowneySr committed:
Merge branch 'carthage' into carthage-amm

dobertRowneySr committed:
fix: build errors for tests

dobertRowneySr committed:
continue on implementing amm activation

dobertRowneySr committed:
fix: run fomatting

dobertRowneySr committed:
chore: add extra activate behavior

dobertRowneySr committed:
feat: add tests for bonding curve parm validation

dobertRowneySr committed:
feat: remove parameters validation for bonding curve

dobertRowneySr committed:
chore: setup BondFixture

dobertRowneySr committed:
feat: add bond tests

dobertRowneySr committed:
feat: add `creator_member_id` to `TokenData`

dobertRowneySr committed:
fix: all test except `timestamp` fixed

dobertRowneySr committed:
fix: remove commented code

dobertRowneySr committed:
fix: test formatting

dobertRowneySr committed:
fix: add timestamp and slippage errors

dobertRowneySr committed:
feat: finish tests for `bond` extrinsic

dobertRowneySr committed:
feat: add unbonding tests

dobertRowneySr committed:
fix: review and correct `unbond` tests

dobertRowneySr committed:
fix: all tests passing

dobertRowneySr committed:
feat: add `deactivate_amm` tests

dobertRowneySr committed:
fix: all tests passing

dobertRowneySr committed:
fix: pricing function computation

dobertRowneySr committed:
fix: all tests passing

dobertRowneySr committed:
fix: quick refactor

dobertRowneySr committed:
fix: slippage calculations and clippy errors

dobertRowneySr committed:
fix: added events

dobertRowneySr committed:
feat: add event deposit and tests

dobertRowneySr committed:
chore: add comments

dobertRowneySr committed:
fix: clippy

dobertRowneySr committed:
fix: all tests passing

dobertRowneySr committed:
fix: benchmark tests

Lezek123 commented on runtime-modules/project-token/src/errors.rs:
```diff
@@ -159,5 +162,34 @@ decl_error! {
+        /// Deadline timestamp lower than now
+        InvalidTimestampSpecified,
+
+        /// Deadline constraint not respected
```
```suggestion
        /// Deadline constraint not satisfied
```

Lezek123 commented on runtime-modules/project-token/src/errors.rs:
```diff
@@ -159,5 +162,34 @@ decl_error! {
+        /// Deadline constraint not respected
+        DeadlineExpired,
+
+        /// SLippage tolerance constraint tolerance not respected
```
```suggestion
        /// Slippage tolerance constraint tolerance not satisfied
```

Lezek123 commented on runtime-modules/project-token/src/errors.rs:
```diff
@@ -159,5 +162,34 @@ decl_error! {
+        SlippageToleranceExceeded,
+
+        /// SLippage tolerance constraint tolerance not respected
+        UnsufficientTokenAmount,
```
```suggestion
        /// Creator token balance is insufficient 
        InsufficientTokenBalance,
```

Lezek123 commented on runtime-modules/project-token/src/events.rs:
```diff
@@ -155,5 +155,34 @@ decl_event! {
+        /// - token id
+        /// - member id
+        /// - params for the bonding curve
+        BondingCurveActivated(TokenId, MemberId, BondingCurve),
```
My general feeling about this naming strategy (both for events and extrinsics) is that it's very difficult to parse for multiple reasons:
- bonding/unbonding may refer to many things and is often used in context of staking (for example, when bonding to nominate/validate).
- It's not clear what you actually `bond`. Is it JOY? But then again, the event is called `TokenBonded`, even though when you call `bond` you actually deposit JOY (and mint the token).

What do you think about using more descriptive names like:
- `AMMActivated`
- `TokensBoughtOnAMM` (you deposit JOY and recieve tokens)
- `TokensSoldOnAMM` (you deposit (burn) tokens and recieve JOY)
- `AMMDeactivated`

(consistently using this naming for extrinsics/tests/constants etc. too)

Lezek123 commented on runtime-modules/project-token/src/lib.rs:
```diff
@@ -800,6 +801,239 @@ decl_module! {
+        /// - amm account created with existential deposit (if necessary)
+        /// - event deposited
+        #[weight = 100_000_000] // TODO: adjust weight
+        fn activate_amm(origin, token_id: T::TokenId, member_id: T::MemberId, curve: BondingCurve) -> DispatchResult {
```
This shouldn't be an extrinsic, instead it should be a method executed by the content pallet, just like `init_token_sale` for example.

The content pallet should perform the authentication, because it must check that the caller is channel owner.

Lezek123 commented on runtime-modules/project-token/src/lib.rs:
```diff
@@ -800,6 +801,239 @@ decl_module! {
+            )?;
+            let token_data = Self::ensure_token_exists(token_id)?;
+
+            ensure!(OfferingStateOf::<T>::ensure_idle_of::<T>(&token_data).is_ok(), Error::<T>::TokenIssuanceNotInIdleState);
```
This is redundant, `ensure_idle_of` already returns `Error::<T>::TokenIssuanceNotInIdleState` on failure:
```suggestion
            OfferingStateOf::<T>::ensure_idle_of::<T>(&token_data)?;
```

Lezek123 commented on runtime-modules/project-token/src/lib.rs:
```diff
@@ -800,6 +801,239 @@ decl_module! {
+            Ok(())
+        }
+
+        /// Activate Amm functionality for the token
```
Invalid comment for `fn bond`

Lezek123 commented on runtime-modules/project-token/src/lib.rs:
```diff
@@ -800,6 +801,239 @@ decl_module! {
+        /// Preconditions
+        /// - origin, member_id pair must be a valid authentication pair
+        /// - token_id must exist
+        /// - user usable JOY balance must be enough for bonding (+ existential deposit eventually)
```
```suggestion
        /// - user usable JOY balance must be enough for bonding (+ existential deposit)
```

Lezek123 commented on runtime-modules/project-token/src/lib.rs:
```diff
@@ -800,6 +801,239 @@ decl_module! {
+        /// - token total supply and amount value must be s.t. `eval` function doesn't overflow
+        ///
+        /// Postconditions
+        /// - token `bonding_curve` activated with specified parameters
```
Invalid postcondition for `fn bond`

Lezek123 commented on runtime-modules/project-token/src/lib.rs:
```diff
@@ -800,6 +801,239 @@ decl_module! {
+        ///
+        /// Postconditions
+        /// - token `bonding_curve` activated with specified parameters
+        /// - `amount` CRT minted into account (which is created if necessary and existential deposit added to it)
```
```suggestion
        /// - `amount` CRT minted into account (which is created if necessary with existential deposit transferred to it)
```

Lezek123 commented on runtime-modules/project-token/src/lib.rs:
```diff
@@ -800,6 +801,239 @@ decl_module! {
+
+            // slippage tolerance check
+            if let Some((slippage_tolerance, desired_price)) = slippage_tolerance {
+                ensure!(price.saturating_sub(desired_price) < slippage_tolerance.mul_floor(desired_price), Error::<T>::SlippageToleranceExceeded);
```
I think this would be more correct
```suggestion
                ensure!(price.saturating_sub(desired_price) <= slippage_tolerance.mul_floor(desired_price), Error::<T>::SlippageToleranceExceeded);
```
(For example, with `price = 100`, `slippage_tolerance = 0.02`, `102` should still be an acceptable price)

Lezek123 commented on runtime-modules/project-token/src/lib.rs:
```diff
@@ -800,6 +801,239 @@ decl_module! {
+
+            // timestamp deadline check
+            if let Some(deadline) = deadline {
+                ensure!(<timestamp::Pallet<T>>::get() < deadline, Error::<T>::DeadlineExpired);
```
```suggestion
                ensure!(<timestamp::Pallet<T>>::get() <= deadline, Error::<T>::DeadlineExpired);
```
(same as with `slippage_tolerance`, for consistency)

Lezek123 commented on runtime-modules/project-token/src/lib.rs:
```diff
@@ -800,6 +801,239 @@ decl_module! {
+
+            // == MUTATION SAFE ==
+
+            let creator_member_id = token_data.creator_member_id;
```
Keeping `creator_member_id` in `token_data` is not a good idea, because the creator token should be effectively owned by the current channel owner (which is subject to change).

If we take _Sale_ as an example, each time a sale is initialized, the content pallet passes `earnings_dst` to `T::ProjectToken::init_token_sale`, and the `earnings_dst` always depends on the current channel owner.

Additionally, **one cannot transfer a channel while creator token sale is ongoing**. A similar mechanism would be needed for AMM (_see `Content::initialize_channel_transfer`_).

This way we can ensure than only the current channel owner profits from a creator token Sale/AMM.

dobertRowneySr replied:
I am thinking whether to simply add transaction fees instead. Since we already have patronage that has a similar mechanism to this. 
I have used this mechanics because Deso works in a similar fashion, but at this point I think that having a platform-wide buy/sell tx % fees would be overall simpler

Lezek123 commented on runtime-modules/project-token/src/lib.rs:
```diff
@@ -800,6 +801,239 @@ decl_module! {
+            let creator_amount = curve.creator_reward.mul_floor(amount);
+            if AccountInfoByTokenAndMember::<T>::contains_key(token_id, creator_member_id) {
+                AccountInfoByTokenAndMember::<T>::mutate(token_id, creator_member_id, |account_data| {
+                    account_data.amount = account_data.amount.saturating_add(creator_amount)
```
```suggestion
                    account_data.increase_amount_by(creator_amount);
```

Lezek123 commented on runtime-modules/project-token/src/lib.rs:
```diff
@@ -800,6 +801,239 @@ decl_module! {
+                });
+            }
+
+            let user_amount = amount - creator_amount;
```
```suggestion
            let user_amount = amount.saturating_sub(creator_amount);
```
That's a shame that `deny(clippy::integer_arithmetic)` doesn't catch that :/
We will probably need to use https://github.com/rust-lang/rust-clippy/pull/9130 in the furute.

Lezek123 commented on runtime-modules/project-token/src/lib.rs:
```diff
@@ -800,6 +801,239 @@ decl_module! {
+                Self::transfer_joy(&sender, &Self::module_treasury_account(), bloat_bond)?;
+            } else {
+                AccountInfoByTokenAndMember::<T>::mutate(token_id, member_id, |account_data| {
+                    account_data.amount = account_data.amount.saturating_add(user_amount);
```
```suggestion
                    account_data.increase_amount_by(user_amount);
```

Lezek123 commented on runtime-modules/project-token/src/lib.rs:
```diff
@@ -800,6 +801,239 @@ decl_module! {
+            Ok(())
+        }
+
+        /// Mint token and transfer joys to the AMM
```
Invalid comment for `fn unbond`

Lezek123 commented on runtime-modules/project-token/src/lib.rs:
```diff
@@ -800,6 +801,239 @@ decl_module! {
+        /// - amm treasury acconut must have sufficient JOYs for the operation
+        ///
+        /// Postconditions
+        /// - `amount` burn from user account
```
```suggestion
        /// - `amount` burned from user account
```

Lezek123 commented on runtime-modules/project-token/src/lib.rs:
```diff
@@ -800,6 +801,239 @@ decl_module! {
+
+            // slippage tolerance check
+            if let Some((slippage_tolerance, desired_price)) = slippage_tolerance {
+                ensure!(desired_price.saturating_sub(price) < slippage_tolerance.mul_floor(desired_price), Error::<T>::SlippageToleranceExceeded);
```
```suggestion
                ensure!(desired_price.saturating_sub(price) <= slippage_tolerance.mul_floor(desired_price), Error::<T>::SlippageToleranceExceeded);
```

Lezek123 commented on runtime-modules/project-token/src/lib.rs:
```diff
@@ -800,6 +801,239 @@ decl_module! {
+
+            // timestamp deadline check
+            if let Some(deadline) = deadline {
+                ensure!(<timestamp::Pallet<T>>::get() < deadline, Error::<T>::DeadlineExpired);
```
```suggestion
                ensure!(<timestamp::Pallet<T>>::get() <= deadline, Error::<T>::DeadlineExpired);
```

Lezek123 commented on runtime-modules/project-token/src/lib.rs:
```diff
@@ -800,6 +801,239 @@ decl_module! {
+            // == MUTATION SAFE ==
+
+            AccountInfoByTokenAndMember::<T>::mutate(token_id, member_id, |account_data| {
+                account_data.amount = account_data.amount.saturating_sub(amount)
```
```suggestion
                account_data.decrease_amount_by(amount);
```

Lezek123 commented on runtime-modules/project-token/src/lib.rs:
```diff
@@ -800,6 +801,239 @@ decl_module! {
+            });
+
+            TokenInfoById::<T>::mutate(token_id, |token_data| {
+                token_data.total_supply = token_data.total_supply.saturating_sub(amount);
```
```suggestion
                token_data.decrease_supply_by(amount);
```

Lezek123 commented on runtime-modules/project-token/src/lib.rs:
```diff
@@ -800,6 +801,239 @@ decl_module! {
+            let curve = token_data.bonding_curve.ok_or(Error::<T>::NotInAmmState)?;
+            let user_acc_data = Self::ensure_account_data_exists(token_id, &member_id)?;
+
+            ensure!(user_acc_data.amount >= amount, Error::<T>::UnsufficientTokenAmount);
```
To prevent vesting funds leakage you should check `transferrable` balance here:
```suggestion
            ensure!(
                user_acc_data.transferrable::<T>(Self::current_block()) >= amount,
                Error::<T>::InsufficientTokenAmount
            );
```

Lezek123 commented on runtime-modules/project-token/src/lib.rs:
```diff
@@ -800,6 +801,239 @@ decl_module! {
+
+            TokenInfoById::<T>::mutate(token_id, |token_data| {
+                token_data.total_supply = token_data.total_supply.saturating_add(amount);
+                token_data.tokens_issued = token_data.tokens_issued.saturating_add(amount);
```
```suggestion
                token_data.increase_supply_by(amount);
```
(already does both)

Lezek123 commented on runtime-modules/project-token/src/lib.rs:
```diff
@@ -800,6 +801,239 @@ decl_module! {
+        /// - state set to idle
+        /// - event deposited
+        #[weight = 10_000_000] // TODO: Adjust weight
+        fn deactivate_amm(origin, token_id: T::TokenId, member_id: T::MemberId) -> DispatchResult {
```
Just as `activate_amm` - this shouldn't be an extrinsic (for the same reason)

dobertRowneySr replied:
I have made them method + added the extrinsic version (with test) on the content pallet

Lezek123 commented on runtime-modules/project-token/src/lib.rs:
```diff
@@ -800,6 +801,239 @@ decl_module! {
+
+            let token_data = Self::ensure_token_exists(token_id)?;
+
+            ensure!(token_data.creator_member_id == member_id, Error::<T>::UserNotAuthorized);
```
This will go away if we shift owner authentication to `content` pallet (which is where it should be done)

Lezek123 commented on runtime-modules/project-token/src/lib.rs:
```diff
@@ -800,6 +801,239 @@ decl_module! {
+
+            ensure!(token_data.creator_member_id == member_id, Error::<T>::UserNotAuthorized);
+
+            ensure!(OfferingStateOf::<T>::ensure_bonding_curve_of::<T>(&token_data).is_ok(), Error::<T>::NotInAmmState);
```
`ensure_bonding_curve_of` already returns `Error::<T>::NotInAmmState` on failure:
```suggestion
            OfferingStateOf::<T>::ensure_bonding_curve_of::<T>(&token_data)?;
```

Lezek123 commented on runtime-modules/project-token/src/lib.rs:
```diff
@@ -800,6 +801,239 @@ decl_module! {
+            ensure!(OfferingStateOf::<T>::ensure_bonding_curve_of::<T>(&token_data).is_ok(), Error::<T>::NotInAmmState);
+
+            let amm_treasury_account = Self::module_bonding_curve_reserve_account(token_id);
+            ensure!(Joy::<T>::usable_balance(amm_treasury_account) == T::JoyExistentialDeposit::get(), Error::<T>::AmmTreasuryBalanceNotEmpty);
```
In order to satisfy this condition, as many tokens as were ever bought on this AMM would need to be sold, which may be impossible. I think there should be some alternative way of closing the AMM (perhaps with some minimum deadline, after which all remaining JOYs are burned?)

dobertRowneySr replied:
I have added an extra variable on the `BondingCurve` data structure that tracks how much supply has been created via the AMM. The condition for deactivating the amm is now:

`(amount_minted_through_amm/total_supply) <= 5%`

The threshold 5% can be fixed by the governance

Lezek123 commented on runtime-modules/project-token/src/lib.rs:
```diff
@@ -1619,6 +1853,11 @@ impl<T: Config> Module<T> {
 
+    /// Returns teh account for the bonding curve treasury
+    pub fn module_bonding_curve_reserve_account(token_id: T::TokenId) -> T::AccountId {
+        <T as Config>::ModuleId::get().into_sub_account_truncating(&("BONDING_RESERVE", token_id))
```
It's risky to use long strings here (longer than 10-12 bytes), because `into_sub_account_truncating` will:
- Add a prefix which is 12 bytes
- Concat it with your input (`BONDING_RESERVE` string itself is already 15 bytes)
- Turncate the result to match `AccountId` length (which is 32 bytes)

If the input is too long, an important part of it (like `token_id` in this case) may get turncated and this will cause huge issues.

How about:
```suggestion
        <T as Config>::ModuleId::get().into_sub_account_truncating(&("AMM", token_id))
```

Lezek123 commented on runtime-modules/project-token/src/types.rs:
```diff
@@ -569,6 +575,57 @@ where
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[derive(Default, Encode, Decode, TypeInfo, Clone, Debug, Eq, PartialEq, MaxEncodedLen)]
+pub struct BondingCurve {
+    /// Slope parameter : a
+    pub slope: Permill,
```
This prevents a slope from being > 1, which is very limiting

dobertRowneySr replied:
I don't really think that would be a problem. If you look at Deso you see that `a=.003`. The square term already makes the result quite large.
But in any case I think this problem can be solved on the next PR which takes care on all the constraints for the AMM

Lezek123 replied:
In the end it all depends on the token supply and decimals choosen by the creator.
If I want my token supply to be 1k and have only 2 decimals, then there will be just 100000 units of my CRT. In this case `a=1` is not a very high value, as buying/selling all 100000 units will only move the price (per unit) by 100000 HAPI, which is almost nothing

Lezek123 commented on runtime-modules/project-token/src/types.rs:
```diff
@@ -569,6 +575,57 @@ where
+    /// Slope parameter : a
+    pub slope: Permill,
+
+    /// Intercept : b
+    pub intercept: Permill,
```
Same problem as with the slope

dobertRowneySr replied:
see reply for the `slope` parameter

Lezek123 replied:
For `b` the problem is even more pronounced, because this is the main parameter which allows controlling the initial price. If it cannot be more than 1, then creator has almost no control over this.

I don't have an issue with this being fixed in the next PR, but I don't agree it's not a problem (this applies to both `a` and `b` actually)

Lezek123 commented on runtime-modules/project-token/src/tests/bonding_curve.rs:
```diff
@@ -0,0 +1,1079 @@
+                Token::ensure_account_data_exists(token_id, &user_member_id).unwrap();
+            let user_amount = BONDING_CURVE_CREATOR_REWARD
+                .left_from_one()
+                .mul_floor(DEFAULT_BONDING_AMOUNT);
```
That's not always going to match `amount - creator_amount`.
For example, imagine `amount = 101` and `BONDING_CURVE_CREATOR_REWARD = 1%`.
Then:
```
creator_amount = 1
amount - creator_amount = 100
(0.99).mul_floor(amount) = 99
```

dobertRowneySr replied:
I removed the creator reward in favor of tx fees, as it seemed more apt to me

Lezek123 commented on runtime-modules/project-token/src/tests/bonding_curve.rs:
```diff
@@ -0,0 +1,1079 @@
+                .amount;
+            let user_amount_increase = BONDING_CURVE_CREATOR_REWARD
+                .left_from_one()
+                .mul_floor(DEFAULT_BONDING_AMOUNT);
```
Same comment as before: this doesn't always match `amount - creator_amount`

dobertRowneySr replied:
I removed the creator reward in favor of tx fees, as it seemed more apt to me

Lezek123 commented on runtime-modules/project-token/src/tests/bonding_curve.rs:
```diff
@@ -0,0 +1,1079 @@
+                .unwrap();
+
+            let amm_reserve_post = Balances::usable_balance(amm_reserve_account);
+            let correctly_computed_joy_amount = 1000500;
```
Some explaination (comment) of why and under what conditions (ie. assuming `DEFAULT_BONDING_AMOUNT=x`) this is the correct amount would be advised

dobertRowneySr replied:
I added an helper function to make the computation clearer

Lezek123 commented on runtime-modules/project-token/src/tests/bonding_curve.rs:
```diff
@@ -0,0 +1,1079 @@
+fn user_joy_balance_correctly_decreased_during_bonding() {
+    let (_, user_account) = member!(2);
+    let (user_account_id, user_balance) = (member!(2).1, joy!(5_000_000));
+    let correctly_computed_joy_amount = 1000500;
```
Same comment as before

dobertRowneySr replied:
I added an helper function to make the computation clearer

Lezek123 commented on runtime-modules/project-token/src/tests/bonding_curve.rs:
```diff
@@ -0,0 +1,1079 @@
+}
+
+#[test]
+fn crt_correctly_minted_to_user_during_bonding() {
```
Isn't this test effectively repeating the assertions that are already part of `bonding_succeeds_with_existing_user` / `bonding_succeeds_with_new_user`?

Lezek123 commented on runtime-modules/project-token/src/tests/bonding_curve.rs:
```diff
@@ -0,0 +1,1079 @@
+fn bond_ok_with_event_deposit() {
+    let token_id = token!(1);
+    let ((user_member_id, user_account_id), user_balance) = (member!(2), joy!(5_000_000));
+    let price = joy!(1000500);
```
Same as before - a comment explaining where `1000500` comes from would be advised

dobertRowneySr replied:
I added an helper function to make the computation clearer

Lezek123 commented on runtime-modules/project-token/src/tests/bonding_curve.rs:
```diff
@@ -0,0 +1,1079 @@
+}
+
+#[test]
+fn unbond_fails_with_user_not_having_enought_crt() {
```
After fixing the vesting leak issue, it'll be a good idea to add test for that too (ie. vesting funds cannot be used to "unbond")

Lezek123 commented on runtime-modules/project-token/src/tests/bonding_curve.rs:
```diff
@@ -0,0 +1,1079 @@
+                .unwrap();
+
+            let amm_reserve_post = Balances::usable_balance(amm_reserve_account);
+            let correctly_computed_joy_amount = 100095;
```
Please add a comment describing where this value comes from

dobertRowneySr replied:
I added an helper function to make the computation clearer

Lezek123 commented on runtime-modules/project-token/src/tests/bonding_curve.rs:
```diff
@@ -0,0 +1,1079 @@
+fn unbonding_ok_with_user_joy_balance_correctly_increased() {
+    let (_, user_account) = member!(2);
+    let (user_account_id, user_balance) = (member!(2).1, joy!(5_000_000));
+    build_default_test_externalities_with_balances(vec![(user_account_id, user_balance)])
```
```suggestion
    let (user_account, user_balance) = (member!(2).1, joy!(5_000_000));
    build_default_test_externalities_with_balances(vec![(user_account, user_balance)])
```

Lezek123 commented on runtime-modules/project-token/src/tests/bonding_curve.rs:
```diff
@@ -0,0 +1,1079 @@
+            UnbondFixture::default().execute_call().unwrap();
+
+            let user_reserve_post = Balances::usable_balance(user_account);
+            let correctly_computed_joy_amount = 95;
```
Please add an explainer (comment) as to where this value comes from

dobertRowneySr replied:
I added an helper function to make the computation clearer

Lezek123 commented on runtime-modules/project-token/src/tests/bonding_curve.rs:
```diff
@@ -0,0 +1,1079 @@
+fn unbonding_ok_with_event_deposited() {
+    let token_id = token!(1);
+    let ((user_id, user_account), user_balance) = (member!(2), joy!(5_000_000));
+    let price = joy!(95); // see last test
```
Please add explainer comment

dobertRowneySr replied:
I have added an helper function with the full explicit formula i.e `a/2 * ((supply + amount)^2 - supply^2) + b * amount`

Lezek123 commented on runtime-modules/project-token/src/tests/bonding_curve.rs:
```diff
@@ -0,0 +1,1079 @@
+}
+
+#[test]
+fn token_state_idle_after_amm_deactivation() {
```
Already checked in `deactivate_ok_with_status_set_to_idle`

Lezek123 commented on runtime-modules/project-token/src/tests/bonding_curve.rs:
```diff
@@ -0,0 +1,1079 @@
+                ..
+            } = Token::token_info_by_id(token_id);
+
+            // .001 * ((1ml + x)^2 - 1ml^2)/2
```
1. I think it's better to use simplified formula `1/2 * ax^2 + asx + bx`, which is also used by the `BondingCurve` type (`s` is supply, `a`, `b` are linear function params, `x` is amount)
2. Ideally the comment would state both the formula and the actual values used, ie.:
```suggestion
            // 1/2 * a * x^2 + a * s * x + b * x = 
            // 1/2 * .001 * x^2 + .001 * 10^6 * x =
            // 0.0005 * x^2 + 1000 * x
```

Lezek123 commented on runtime-modules/project-token/src/tests/bonding_curve.rs:
```diff
@@ -0,0 +1,1079 @@
+// --------------------------------- EXTRA ----------------------------------------
+
+#[test]
+fn review_eval_function() {
```
Also, I think you should use some non-zero `b` for this test, otherwise it seems kind of incomplete

Lezek123 commented on runtime-modules/project-token/src/tests/bonding_curve.rs:
```diff
@@ -0,0 +1,1079 @@
+                })
+                .collect::<Vec<_>>();
+
+            // -[.001 * ((1ml - x)^2 - 1ml^2)/2]
```
I suggest expanding this comment in a similar fashion to how I did it for the previous one

dobertRowneySr replied:
I have added an helper function to make things clearer

Commit:
Update runtime-modules/project-token/src/errors.rs

Co-authored-by: Leszek Wiesner <leszek@jsgenesis.com>

Commit:
Update runtime-modules/project-token/src/errors.rs

Co-authored-by: Leszek Wiesner <leszek@jsgenesis.com>

Commit:
Update runtime-modules/project-token/src/errors.rs

Co-authored-by: Leszek Wiesner <leszek@jsgenesis.com>

Commit:
Update runtime-modules/project-token/src/tests/bonding_curve.rs

Co-authored-by: Leszek Wiesner <leszek@jsgenesis.com>

Commit:
Update runtime-modules/project-token/src/lib.rs

Co-authored-by: Leszek Wiesner <leszek@jsgenesis.com>

Commit:
Update runtime-modules/project-token/src/lib.rs

Co-authored-by: Leszek Wiesner <leszek@jsgenesis.com>

Commit:
Update runtime-modules/project-token/src/lib.rs

Co-authored-by: Leszek Wiesner <leszek@jsgenesis.com>

Commit:
Update runtime-modules/project-token/src/lib.rs

Co-authored-by: Leszek Wiesner <leszek@jsgenesis.com>

Commit:
Update runtime-modules/project-token/src/lib.rs

Co-authored-by: Leszek Wiesner <leszek@jsgenesis.com>

Commit:
Update runtime-modules/project-token/src/lib.rs

Co-authored-by: Leszek Wiesner <leszek@jsgenesis.com>

Commit:
Update runtime-modules/project-token/src/lib.rs

Co-authored-by: Leszek Wiesner <leszek@jsgenesis.com>

Commit:
Update runtime-modules/project-token/src/lib.rs

Co-authored-by: Leszek Wiesner <leszek@jsgenesis.com>

Commit:
Update runtime-modules/project-token/src/lib.rs

Co-authored-by: Leszek Wiesner <leszek@jsgenesis.com>

Commit:
Update runtime-modules/project-token/src/lib.rs

Co-authored-by: Leszek Wiesner <leszek@jsgenesis.com>

Commit:
Update runtime-modules/project-token/src/lib.rs

Co-authored-by: Leszek Wiesner <leszek@jsgenesis.com>

Commit:
Update runtime-modules/project-token/src/lib.rs

Co-authored-by: Leszek Wiesner <leszek@jsgenesis.com>

Commit:
Update runtime-modules/project-token/src/lib.rs

Co-authored-by: Leszek Wiesner <leszek@jsgenesis.com>

Commit:
Update runtime-modules/project-token/src/lib.rs

Co-authored-by: Leszek Wiesner <leszek@jsgenesis.com>

Commit:
Update runtime-modules/project-token/src/lib.rs

Co-authored-by: Leszek Wiesner <leszek@jsgenesis.com>

Commit:
Update runtime-modules/project-token/src/lib.rs

Co-authored-by: Leszek Wiesner <leszek@jsgenesis.com>

Commit:
Update runtime-modules/project-token/src/lib.rs

Co-authored-by: Leszek Wiesner <leszek@jsgenesis.com>

Commit:
Update runtime-modules/project-token/src/tests/bonding_curve.rs

Co-authored-by: Leszek Wiesner <leszek@jsgenesis.com>

dobertRowneySr committed:
fix: new mechanism for deactivation amm

dobertRowneySr committed:
fix: add correct value check

dobertRowneySr committed:
fix: activate / deactivate now are methods

dobertRowneySr committed:
fix: deactivate amm + timestamp tests

dobertRowneySr committed:
feat: add tx fees logic

dobertRowneySr committed:
feat: add content related extrinsics for amm

dobertRowneySr committed:
fix: misc + renamed events

dobertRowneySr committed:
fix: cargo checks

dobertRowneySr commented:
### Extra addition
- I removed the creator token reward in favor of tx fees %
- I added `activate_amm` and `deactivate_amm` extrinsics to the content token pallet, they cannot be invoked when a channel transfer is taking place

dobertRowneySr committed:
fix: AmmControl permission

dobertRowneySr committed:
fix: update token data on benchmarks

bedeho commented:
> I removed the creator token reward in favor of tx fees %

What does this mean?

Lezek123 commented on runtime-modules/project-token/src/lib.rs:
```diff
@@ -1348,6 +1519,78 @@ impl<T: Config>
+
+    /// Activate Amm functionality for the token
+    /// Preconditions
+    /// - (origin, member_id) must be a valid authentication pair
```
No longer applies, since auth is handled by content pallet
```suggestion
```

Lezek123 commented on runtime-modules/project-token/src/errors.rs:
```diff
@@ -159,5 +162,31 @@ decl_error! {
+        OutstandingBondedAmountTooLarge,
+
+        /// Generic user not authorized error
+        UserNotAuthorized,
```
Unused now
```suggestion
```

Lezek123 commented on runtime-modules/project-token/src/lib.rs:
```diff
@@ -163,6 +164,15 @@ decl_storage! { generate_storage_info
         /// Platform fee (percentage) charged on top of each sale purchase (in JOY) and burned
         pub SalePlatformFee get(fn sale_platform_fee) config(): Permill;
+
+        /// Percentage threshold for deactivating the amm functionality
+        pub AmmDeactivationThreshold get(fn amm_deactivation_threshold) config(): Permill;
```
I'm not sure what exact problem is this solving.

Suppose the channel owner starts the AMM and then the channel popularity increases significantly, which causes the organic growth of the CRT value. In this case it will still be almost impossible to ever close the AMM, unless someone is willing to sell the CRT way below the market value.

Even more extreme example is someone using AMM as initial strategy of distributing the tokens. In that case the percentage of supply minted through an AMM will always be 100%, so all tokens would need to be sold in order for the AMM to be closed.

I think the main reason for limiting the owner's ability to close the AMM in that people may have expectations about market's availability when taking certain positions, ie. some people will just buy CRT only to sell them later. The most obvious way I see to solve this is to force a proper heads up before the market is closed (like 2 weeks for example), by introducing a "gracing period" for AMM closure.

dobertRowneySr replied:
This is the intended behavior.
The CC cannot close the AMM as he pleases. If the public buys 100 CRT through the AMM then the CC cannot close the AMM unless the public has sold back to the AMM at least 95 CRT, this is the intended behavior

Now, the question is: do we want to allow the CC to close the AMM at his discretion, maybe say allowing for a gracing period as suggested (to prevent CC to immediately lock out CRT holders from selling back their token)? @bedeho

Lezek123 commented on runtime-modules/project-token/src/errors.rs:
```diff
@@ -159,5 +162,31 @@ decl_error! {
+        InsufficientTokenBalance,
+
+        /// Amm treasury balance is not zero
+        OutstandingBondedAmountTooLarge,
```
The comment does not match the implementation anymore, but as I mentioned before, I think the implementation should actually be changed

Lezek123 commented on runtime-modules/project-token/src/types.rs:
```diff
@@ -640,6 +733,9 @@ pub struct TokenAllocation<Balance, VestingScheduleParams> {
 pub struct TokenIssuanceParameters<Hash, TokenAllocation, TransferPolicyParams, MemberId: Ord> {
+    /// Creator Member id
+    pub creator_member_id: MemberId,
+
```
Should be removed, there is no longer a use-case for this.
Also the benchmark checks fail: https://github.com/Joystream/joystream/actions/runs/3513430675/jobs/5886244315

Lezek123 commented on runtime-modules/project-token/src/types.rs:
```diff
@@ -569,9 +572,86 @@ where
+
+    pub(crate) fn increase_bonded_amount_by(&mut self, amount: Balance) {
+        self.amount_minted = self.amount_minted.saturating_add(amount);
+    }
```
As I mentioned in the previous review, this naming is very confusing, I didn't just mean the events:
- What is actually being "bonded" is JOY, but here you're referring to `bonded_amount` as if it's the CRT
- You then add the "bonded_amount" to `amount_minted`, as if the terms `bonded` and `minted` are interchangeable, but they actually mean two completely different things

As I mentioned before, I think buy/sell terminology is much simpler, but if you want to stick with bond/unbond, it should be specified what is actually being bonded and unbonded, because in the current implementation it's JOY, not the CRT, ie.:
- You bond JOY, you mint CRT
- You burn CRT, you unbond JOY

dobertRowneySr replied:
see e257c40d78619458146c9281225b6d566f6450d9 for details

Lezek123 commented on runtime-modules/project-token/src/lib.rs:
```diff
@@ -800,6 +810,166 @@ decl_module! {
+                token_data.increase_bonded_amount_by(amount);
+            });
+
+            // TODO: redirect tx fees revenue to council
```
Was it already decided that the tx fees should go to the council?
Some of the alternatives include:
- Burning (just like in case of CRT sale fees)
- Fees going to channel owner

dobertRowneySr replied:
The destination from the proceedings from TX fees is not established yet.

dobertRowneySr replied:
I will leave the TODO as it is as I am creating an issue with all open questions for AMM to be addressed

Lezek123 commented on runtime-modules/project-token/src/lib.rs:
```diff
@@ -800,6 +810,166 @@ decl_module! {
+                ensure!(<timestamp::Pallet<T>>::now() <= deadline, Error::<T>::DeadlineExpired);
+            }
+
+            let unbond_price = Self::unbond_tx_fees().left_from_one().mul_floor(price);
```
This is actually more like an `unbonded_amount` (which comes back to the naming issue)

dobertRowneySr replied:
see e257c40d78619458146c9281225b6d566f6450d9 for details

Lezek123 commented on runtime-modules/content/src/lib.rs:
```diff
@@ -3748,6 +3749,72 @@ decl_module! {
+        ) {
+            let channel = Self::ensure_channel_exists(&channel_id)?;
+
+            channel.ensure_has_no_active_transfer::<T>()?;
```
I don't think this check is needed, instead you should check whether there is an active AMM during `initialize_transfer` in the first place and disallow the transfer in this case.

See: https://github.com/Joystream/joystream/blob/mainnet/runtime-modules/content/src/lib.rs#L3061, we're already ensuring there is no active revenue split and/or active/scheduled sale, adding the same condition for AMM would make sense.

Lezek123 commented on runtime-modules/project-token/src/lib.rs:
```diff
@@ -1348,6 +1519,78 @@ impl<T: Config>
+        let amm_treasury_account = Self::module_bonding_curve_reserve_account(token_id);
+        let amount_to_slash = Joy::<T>::usable_balance(&amm_treasury_account)
+            .saturating_sub(T::JoyExistentialDeposit::get());
+        let _ = Joy::<T>::slash(&amm_treasury_account, amount_to_slash);
```
It's not really and issue here, but I'd still advise using a fallible `burn_from_usable` instead, to make sure the entire provided amount is always burned

Lezek123 commented on runtime-modules/project-token/src/lib.rs:
```diff
@@ -1348,6 +1519,78 @@ impl<T: Config>
+            .saturating_sub(T::JoyExistentialDeposit::get());
+        let _ = Joy::<T>::slash(&amm_treasury_account, amount_to_slash);
+
+        Self::deposit_event(RawEvent::AmmDeactivated(token_id, member_id));
```
I think burned amount would be a useful information to include in this event

Lezek123 commented on runtime-modules/project-token/src/tests/bonding_curve.rs:
```diff
@@ -0,0 +1,924 @@
+
+        // ----------------- act -------------------
+        let result = UnbondFixture::default()
+            .with_amount(DEFAULT_BONDING_AMOUNT)
```
Alice only purchased `DEFAULT_UNBONDING_AMOUNT` (100) tokens, but you're trying to burn `DEFAULT_BONDING_AMOUNT` which is 1000, so the test doesn't really check what it's supposed to check.

Notice that `DEFAULT_UNBONDING_AMOUNT` is also kind of a confusing term, because what is getting unbonded is JOY, and the CRT would be in this case a "burned amount".

Another suggestion: Please use `with_sender(alice_account).with_member_id(alice_id)`, I know those are the default values, but it was by no means clear for me until I checked the `UnbondFixture` implementation.

dobertRowneySr replied:
that was a typo on my end, good catch

Lezek123 commented on runtime-modules/project-token/src/tests/bonding_curve.rs:
```diff
@@ -0,0 +1,924 @@
+                .execute_call();
+
+            // ---------------- assert -----------------
+            // Alice is now being vested but she has 0 transferrable amount
```
This comment doesn't apply here

Lezek123 commented on runtime-modules/project-token/src/lib.rs:
```diff
@@ -800,6 +810,166 @@ decl_module! {
+        /// Postconditions
+        /// - `amount` CRT minted into account (which is created if necessary with existential deposit transferred to it)
+        /// - respective JOY amount transferred from user balance to amm treasury account
+        /// - percentage of the minted CRT transferred to the token content creator (creator reward)
```
No longer applies

Lezek123 commented on runtime-modules/project-token/src/lib.rs:
```diff
@@ -800,6 +810,166 @@ decl_module! {
+        /// - deadline constraint respected if provided
+        /// - slippage tolerance constraints respected if provided
+        /// - token total supply and amount value must be s.t. `eval` function doesn't overflow
+        /// - amm treasury acconut must have sufficient JOYs for the operation
```
```suggestion
        /// - amm treasury account must have sufficient JOYs for the operation
```

Commit:
Update runtime-modules/project-token/src/lib.rs

Co-authored-by: Leszek Wiesner <leszek@jsgenesis.com>

Commit:
Update runtime-modules/project-token/src/errors.rs

Co-authored-by: Leszek Wiesner <leszek@jsgenesis.com>

Commit:
Update runtime-modules/project-token/src/lib.rs

Co-authored-by: Leszek Wiesner <leszek@jsgenesis.com>

dobertRowneySr committed:
fix: misc fixes

dobertRowneySr committed:
fix: renaming

bond -> amm buy
unbond -> amm sell
bonding curve -> amm
bonding reserve account -> amm treasury account
amount minted by amm -> amount bought on amm

dobertRowneySr committed:
chore: generate types and metadata

dobertRowneySr committed:
chore: generate types and metadata

dobertRowneySr committed:
Merge branch 'mainnet' into carthage-amm

Lezek123 commented on runtime-modules/content/src/lib.rs:
```diff
@@ -3760,8 +3765,6 @@ decl_module! {
             let channel = Self::ensure_channel_exists(&channel_id)?;
 
-            channel.ensure_has_no_active_transfer::<T>()?;
-
```
In case of `activate_amm` we still need this to prevent someone from starting transfer and then activating AMM before it's finalized (for `deactivate_amm` it's not needed, because you already ensured amm cannot be active when initializing a transfer)

Lezek123 commented on runtime-modules/project-token/src/lib.rs:
```diff
@@ -1351,6 +1530,81 @@ impl<T: Config>
+    /// Preconditions
+    /// - (origin, member_id) must be a valid authentication pair
+    /// - token_id must be a valid
+    /// - token must be in `BondigCurve` state
```
```suggestion
    /// - token must be in `Amm` state
```

Lezek123 commented on runtime-modules/project-token/src/lib.rs:
```diff
@@ -1351,6 +1530,81 @@ impl<T: Config>
+    /// - token must be in `BondigCurve` state
+    ///
+    /// Postconditions
+    /// - Bonding Curve set to None
```
```suggestion
    /// - Amm Curve set to None
```

Lezek123 commented on runtime-modules/project-token/src/lib.rs:
```diff
@@ -1622,6 +1876,11 @@ impl<T: Config> Module<T> {
         <T as Config>::ModuleId::get().into_sub_account_truncating(Vec::<u8>::new())
     }
 
+    /// Returns teh account for the bonding curve treasury
```
```suggestion
    /// Returns the account for the AMM treasury
```

Lezek123 commented on runtime-modules/project-token/src/tests/amm.rs:
```diff
@@ -0,0 +1,919 @@
+}
+
+#[test]
+fn bond_ok_with_event_deposit() {
```
```suggestion
fn amm_buy_ok_with_event_deposit() {
```

Lezek123 commented on runtime-modules/project-token/src/tests/amm.rs:
```diff
@@ -0,0 +1,919 @@
+}
+
+#[test]
+fn unbond_fails_with_user_not_having_leaking_funds_from_vesting_schedule() {
```
```suggestion
fn amm_sell_fails_with_user_not_having_leaking_funds_from_vesting_schedule() {
```

Lezek123 commented on runtime-modules/project-token/src/tests/amm.rs:
```diff
@@ -0,0 +1,919 @@
+    })
+}
+#[test]
+fn unbond_fails_with_user_not_having_enough_token_balance() {
```
```suggestion
fn amm_sell_fails_with_user_not_having_enough_token_balance() {
```

Lezek123 commented on runtime-modules/project-token/src/errors.rs:
```diff
@@ -159,5 +162,28 @@ decl_error! {
+        InsufficientTokenBalance,
+
+        /// Outstanding amount of CRT bought from AMM too large
+        OutstandingBondedAmountTooLarge,
```
```suggestion
        /// Oustanding AMM-provided supply constitutes too large percentage of the token's total supply
        OutstandingAmmProvidedSupplyTooLarge,
```

Lezek123 commented on runtime-modules/project-token/src/types.rs:
```diff
@@ -569,9 +572,86 @@ where
+    pub intercept: Permill,
+
+    // amount public bought on amm so far
+    pub amount_bought_on_amm: Balance,
```
```suggestion
    // amount of token added to circulation by the AMM so far
    pub provided_supply: Balance,
```

Lezek123 commented on runtime-modules/project-token/src/events.rs:
```diff
@@ -155,5 +156,35 @@ decl_event! {
+        /// - params for the bonding curve
+        AmmActivated(TokenId, MemberId, AmmCurve),
+
+        /// Token Bonded
```
```suggestion
        /// Tokens Bought on AMM
```

Lezek123 commented on runtime-modules/project-token/src/events.rs:
```diff
@@ -155,5 +156,35 @@ decl_event! {
+        /// - amount of JOY deposited into curve treasury
+        TokensBoughtOnAmm(TokenId, MemberId, Balance, JoyBalance),
+
+        /// Token Unbonded
```
```suggestion
        /// Tokens Sold on AMM
```

Lezek123 commented on runtime-modules/project-token/src/lib.rs:
```diff
@@ -166,6 +167,15 @@ decl_storage! { generate_storage_info
+        /// Percentage threshold for deactivating the amm functionality
+        pub AmmDeactivationThreshold get(fn amm_deactivation_threshold) config(): Permill;
+
+        /// Bond Transaction fee percentage
```
```suggestion
        /// AMM buy transaction fee percentage
```

Lezek123 commented on runtime-modules/project-token/src/lib.rs:
```diff
@@ -166,6 +167,15 @@ decl_storage! { generate_storage_info
+        /// Bond Transaction fee percentage
+        pub AmmBuyTxFees get(fn amm_buy_tx_fees) config(): Permill;
+
+        /// Bond Transaction fee percentage
```
```suggestion
        /// AMM sell transaction fee percentage
```

Lezek123 commented on runtime-modules/project-token/src/tests/amm.rs:
```diff
@@ -0,0 +1,919 @@
+}
+
+#[test]
+fn deactivate_fails_with_too_much_bonded_token_outstanding() {
```
```suggestion
fn deactivate_fails_with_too_much_amm_provided_supply_outstanding() {
```

Lezek123 reviewed (changes requested):
1. One issue with transfers logic
2. A few more renaming suggestions
3. Content benchmarks are failing: https://github.com/Joystream/joystream/actions/runs/3532148688/jobs/5926148030

Commit:
Update runtime-modules/project-token/src/lib.rs

Co-authored-by: Leszek Wiesner <leszek@jsgenesis.com>

Commit:
Update runtime-modules/project-token/src/lib.rs

Co-authored-by: Leszek Wiesner <leszek@jsgenesis.com>

Commit:
Update runtime-modules/project-token/src/events.rs

Co-authored-by: Leszek Wiesner <leszek@jsgenesis.com>

Commit:
Update runtime-modules/project-token/src/lib.rs

Co-authored-by: Leszek Wiesner <leszek@jsgenesis.com>

Commit:
Update runtime-modules/project-token/src/tests/amm.rs

Co-authored-by: Leszek Wiesner <leszek@jsgenesis.com>

Commit:
Update runtime-modules/project-token/src/lib.rs

Co-authored-by: Leszek Wiesner <leszek@jsgenesis.com>

Commit:
Update runtime-modules/project-token/src/lib.rs

Co-authored-by: Leszek Wiesner <leszek@jsgenesis.com>

Commit:
Update runtime-modules/project-token/src/tests/amm.rs

Co-authored-by: Leszek Wiesner <leszek@jsgenesis.com>

Commit:
Update runtime-modules/project-token/src/events.rs

Co-authored-by: Leszek Wiesner <leszek@jsgenesis.com>

Commit:
Update runtime-modules/project-token/src/tests/amm.rs

Co-authored-by: Leszek Wiesner <leszek@jsgenesis.com>

Commit:
Update runtime-modules/project-token/src/tests/amm.rs

Co-authored-by: Leszek Wiesner <leszek@jsgenesis.com>

Commit:
Update runtime-modules/project-token/src/errors.rs

Co-authored-by: Leszek Wiesner <leszek@jsgenesis.com>

Commit:
Update runtime-modules/project-token/src/types.rs

Co-authored-by: Leszek Wiesner <leszek@jsgenesis.com>

dobertRowneySr committed:
fix: transfer logic fix

dobertRowneySr committed:
Merge branch 'mainnet' into carthage-amm

dobertRowneySr committed:
fix: amm curve on BM

Lezek123 commented on runtime-modules/project-token/src/errors.rs:
```diff
@@ -159,5 +162,28 @@ decl_error! {
+        InvalidCurveParameters,
+
+        /// Deadline timestamp lower than now
+        InvalidTimestampSpecified,
```
Unused (also I don't think there's any use-case for this error)

Lezek123 commented on runtime-modules/project-token/src/lib.rs:
```diff
@@ -803,6 +813,165 @@ decl_module! {
+
+            let user_account_data_exists = AccountInfoByTokenAndMember::<T>::contains_key(token_id, &member_id);
+            let amm_treasury_account = Self::amm_treasury_account(token_id);
+            let price = curve.eval::<T>(amount, token_data.total_supply, AmmOperation::Buy)?;
```
The implementation works as expected, I think the only remaining question is what do we want the `supply` parameter in the pricing function to be. There are 2 approaches:
1. We can use actual token's total supply (which can come from sale, patronage etc. and is also affected by non-amm burning)
2. We can use supply provided by the AMM only

I think `2.` would be better, because in this case we don't have to worry about non-amm burns and the patronage affecting the price on the AMM.

### What happens if we use `1.`

If we use token's total supply then the price on the AMM raises when the creator claims the patronage, but the JOY reserve remains unaffected, causing the minimum sell price to raise aswell.

If someone burns tokens outside of the AMM, it's the opposite situation - the price drops, but no JOY is being paid to whoever burned the tokens, which means someone else would eventually claim this JOY from the AMM. In this case the minimum price also drops, making it either harder or impossible to ever close the AMM. Another caveat of non-amm burns is: if we were to allow negative `b`, then it would be possible for the price to drop below 0, meaning someone would get paid for buying CRT (otherwise the AMM would have some JOY locked forever). On the other hand if we don't allow negative `b`, we're limiting the slope quite substantially assuming there were already some tokens sold before the AMM started (for example, via a token sale).

### What happens if we use `2.`

If we were to go with `2.`, everything becomes much more clear. The `b` parameter of the curve is both the initial and the minimum price and it never changes. `a` can be anything `>0`, which means the curve can be anything from almost horizontal line to almost vertical line, giving issuer a lot of freedom when deciding which slope works best for them. We can enforce `b` to be `>= last_sale_price`, which means no AMM can be started that would allow the issuer to dilute the CRT by buying it at a lower price than other members had to pay before and that's it!

dobertRowneySr replied:
This is an old debate.
When it came up I suggested 2 but Bedeho expresses his desire for 1. That's why it is implemented like this
I am in favour of 2 because in that case `b` has the meaning of "price for the first token bought on the amm".
This can (and should be enforced) allow to set `b=last_sale_price`

dobertRowneySr replied:
I passed the `provided_supply` as an argument.
I will refactor the function as part of the next AMM PR, with all the constraints

Lezek123 commented on runtime-modules/project-token/src/tests/fixtures.rs:
```diff
@@ -111,6 +111,47 @@ impl IssueTokenFixture {
+
+    pub fn with_empty_allocation(self) -> Self {
+        self.with_allocation(member!(1).0, 0u64.into())
+    }
```
I tried adding some tests locally and I was confused why the allocation is `> 0` even when `with_empty_allocation` is used. Then I realized `with_allocation` only adds to existing allocation. How about making this method actually clear any existing allocation:
```suggestion
    pub fn with_empty_allocation(self) -> Self {
        Self {
            params: IssuanceParams {
                initial_allocation: BTreeMap::default(),
                ..self.params
            },
            ..self
        }
    }
```

Commit:
Update runtime-modules/project-token/src/tests/fixtures.rs

Co-authored-by: Leszek Wiesner <leszek@jsgenesis.com>

dobertRowneySr committed:
fix: eval function logic edit

eval function now uses amm provided supply of CRT instead of the total
supply amount

dobertRowneySr committed:
fix: remove unused error

Lezek123 reviewed (approved):
LGTM, I'm not sure what the conclusion will be in https://github.com/Joystream/joystream/issues/4151, but we can always adjust in the next PR.

dobertRowneySr commented:
Merging into new branch `crt_release`
